src/sidebar/test/service-url-test.js

```diff
@@ -1,20 +1,166 @@
 'use strict';
 
-var serviceUrl = require('../service-url');
+var proxyquire = require('proxyquire');
 
-describe('serviceUrl', function () {
-  var service;
+/** Return a fake annotationUI object. */
+function fakeAnnotationUI() {
+  var links = null;
+  return {
+    updateLinks: function(newLinks) {
+      links = newLinks;
+    },
+    getState: function() {
+      return {links: links};
+    },
+  };
+}
 
-  beforeEach(function () {
-    service = serviceUrl({serviceUrl: 'https://test.hypothes.is/'});
+function createServiceUrl(linksPromise) {
+  var replaceURLParams = sinon.stub().returns(
+    {url: 'EXPANDED_URL', params: {}}
+  );
```

Comment:
> the mock's outputs don't depend on the inputs, which creates additional ways for the code 
 > under test to do the wrong thing but yet still pass the test.

I don't think this is true (I don't think you _can_ actually change `serviceUrl` so that it calls `replaceURLParams` incorrectly, or does the wrong thing with the return value from `replaceURLParams`, and not have one of my `serviceUrl` tests fail).

I'm not sure that the test here does lose any value as documentation. All `serviceUrl` really does is call `replaceURLParams` and return `replaceURLParams`'s return value. The return value from `replaceURLParams` is an opaque value, as far as `serviceUrl` is concerned. The tests, which are tests for `serviceURL` only after all (`replaceURLParams` has its own tests elsewhere), do document the intended behaviour of `serviceURL` correctly (that it just blindly returns what `replaceURLParams` returns).

I suppose I could have used `"https://example.com/expanded_url"` instead of `"EXPANDED_URL"` but I'm not sure there's much difference (this is another case where I would rather have used `mock.sentinel.expanded_url` if we had that in JavaScript).

There is of course a trade-off that, as you point out, while isolating the `serviceUrl` tests mean that the `serviceUrl` tests won't fail if there's a bug in `replaceURLParams`, if the stub `replaceURLParams` in `service-url-test.js` is wrong or out of date then the `serviceUrl` tests could be passing even though the code is wrong and would crash or misbehave in production. As you point out integrating `service-url-test.js` with the real `replaceURLParams` would avoid this problem.

I fall down on the side of isolating the tests here, rather than integrating them, for a couple of reasons.

One is that isolated tests provide feedback on the design of your code by revealing how deeply and complexly the module under test is coupled to its dependencies. If code is tightly coupled you will end up having to create complex mocks. I think that a couple of the mocks I had to create here to write isolated tests for `serviceUrl` are showing us small examples of unnecessarily tight coupling, which the previous integrated tests for `serviceUrl` didn't reveal.

The second reason is that it scales better. Across a large test suite when you make a lot of small decisions to integrate the tests for a lot of modules with a lot of their dependencies, then you end up in a situation where running the entire test suite is slow, where a bug in one part of the code causes a cascade of test failures all over the test suite making debugging difficult, and where the design of the code in terms of the coupling, interfaces and contracts of modules is bad because you haven't been getting design feedback from isolated tests.